OOP Killer: Added OOP killer to handle Ubuntu package builder

leemgs opened the pull request:
Fixes issue #267.

This commit is to append an OOP killer for Ubuntu builder.

**Changes proposed in this PR:**
1. Added an OOP killer to kill an Ubuntu package builder

Signed-off-by: Geunsik Lim <geunsik.lim@samsung.com>


---

leemgs commented:
Version 1 is not final version. It's just draft version. I mean that we have to discuss enough until making the final version.

leemgs commented:
**Changes proposed in this PR:**
* Version 2:
1. Updated the existing Ubuntu controller based on the module OOP killer
2. Added annotation for code maintenance

* Version 1:
1. Added an OOP killer to kill an Ubuntu package builder

leemgs commented:
To Reviewers, 

Could you give me requirements to complete this PR? 
Your suggestion and advice will be added the "**Action items**" section above.

ohsewon commented on ci/taos/common/out-of-pr-killer.sh:
```diff
@@ -42,19 +43,45 @@ function run_oop_killer(){
+    # The OOP killer for Ubuntu build (pdebuild)
+    # Kill PRs that were previously invoked by checker-pr-gateway.sh with the same PR number.
+    echo "[DEBUG] OOP Killer/Ubuntu: Starting killing activity to kill previously invoked checker-pr-gateway.sh with the same PR number.\n"
+    ps aux | grep "^www-data.*bash \./checker-pr-gateway.sh" | while read line
```
This point is next of tizen oop, so there may be no '/checker-pr-gateway.sh '

leemgs replied:
Good point.  It is no problem.  The duplicated statements does not effect any issue to system although the Tizen OOP killer executed the same statement beforehand. I intend to consider various use cases to support a portability and compatibility.  For example,  please think about the case that a CI administrator try to enable the Ubuntu build module only to compile their GitHub repository.

leemgs commented:
### How to find a pdebuild task of a PR?

![image](https://user-images.githubusercontent.com/82404/48889384-0e1c7680-ee79-11e8-8af3-50d2d95d05fb.png)

leemgs commented:
Fixes issue #267.

This commit is to update the existing Out-of-PR (OOP) killer to control
the Ubuntu build as well as the Tizen build when the PR hang situation happens.

**Changes proposed in this PR:**
* Version 3:
1. Replaced python with bash in order for victim 3 step.

* Version 2:
1. Updated the existing Ubuntu controller based on the module OOP killer
2. Added annotation for code maintenance

* Version 1:
1. Added an OOP killer to kill an Ubuntu package builder

Signed-off-by: Geunsik Lim <geunsik.lim@samsung.com>

myungjoo commented:
My suggestion is to add a PR-number postfix to package name or version/release in debian/control before running pdebuild so that you can identify it easily.

leemgs commented:
**Changes proposed in this PR:**

* Version 4:
1. Added --logfile option to specify an unique name (e.g., PR number, PR time)

* Version 3:
1. Replaced python with bash in order for victim 3 step.

* Version 2:
1. Updated the existing Ubuntu controller based on the module OOP killer
2. Added annotation for code maintenance

* Version 1:
1. Added an OOP killer to kill an Ubuntu package builder

leemgs commented:
**Changes proposed in this PR:**

* Version 5:
1. Fixed incorrect opeartion of Ubuntu/OOP killer
2. Unified the exsting separated two OOP killers (e.g., for Tien, for Ubuntu)

* Version 4:
1. Added --logfile option to specify an unique name (e.g., PR number, PR time)
2. Updated the existing out-of-date comments
3. Specified dependent commands for OOP Killer

* Version 3:
1. Replaced python with bash in order for victim 3 step.

* Version 2:
1. Updated the existing Ubuntu controller based on the module OOP killer
2. Added annotation for code maintenance

* Version 1:
1. Added an OOP killer to kill an Ubuntu package builder

ohsewon commented on ci/taos/common/out-of-pr-killer.sh:
```diff
@@ -27,34 +27,57 @@
+    check_dependency sleep
+    check_dependency rm
+
+    # The OOP killer for Tizen build (gbs)
```
'Tizen build' seems inappropriate.

ohsewon commented on ci/taos/common/out-of-pr-killer.sh:
```diff
@@ -27,34 +27,57 @@
+    # The OOP killer for Tizen build (gbs)
     # Kill PRs that were previously invoked by checker-pr-gateway.sh with the same PR number.
-    echo "[DEBUG] Starting killing activity to kill previously invoked checker-pr-gateway.sh with the same PR number.\n"
+    echo "[DEBUG] OOP Killer/Tizen: Starting killing activity to kill previously invoked checker-pr-gateway.sh with the same PR number.\n"
```
'Tizen' seems inappropriate.

ohsewon commented on ci/taos/common/out-of-pr-killer.sh:
```diff
@@ -27,34 +27,57 @@
+            kill $victim_pid3_tizen
+
+            echo "[DEBUG] OOP Killer/Ubuntu: victim_pr=$victim_pr, input_pr=$input_pr, victim_date=$victim_date, input_date=$input_date "
+            echo "[DEBUG] OOP Killer/Tizen: killing PR $victim_pr (pid <$victim_pid1> <$victim_pid2> <$victim_pid3_ubuntu>)."
```
'Tizen' seems inappropriate.

ohsewon commented on ci/taos/common/out-of-pr-killer.sh:
```diff
@@ -27,34 +27,57 @@
             sleep 1
             # Handle a possibility that someone updates a single PR multiple times within 1 second.
-            echo "[DEBUG] removing the ./${dir_worker}/${victim_pr}-${victim_date}-* folder"
+            echo "[DEBUG] OOP Killer/Tizen: removing the ./${dir_worker}/${victim_pr}-${victim_date}-* folder"
```
'Tizen' seems inappropriate.

leemgs commented:
I unchecked WIP and Do-Not-Merge flag because I have completed local test to avoid unexpected issue that can be generated after merging this PR.  :smile:

leemgs commented:
### Changes proposed in this PR:

* Version 6:
1. Update PR monitor for a ubuntu build
2. Replaced criteria to kill a ubuntu build process (from pbuilder to pdebuild)

* Version 5:
1. Fixed incorrect opeartion of Ubuntu/OOP killer
2. Unified the exsting separated two OOP killers (e.g., for Tien, for Ubuntu)

* Version 4:
1. Added --logfile option to specify an unique name (e.g., PR number, PR time)
2. Updated the existing out-of-date comments
3. Specified dependent commands for OOP Killer

* Version 3:
1. Replaced python with bash in order for victim 3 step.

* Version 2:
1. Updated the existing Ubuntu controller based on the module OOP killer
2. Added annotation for code maintenance

* Version 1:
1. Added an OOP killer to kill an Ubuntu package builder

leemgs commented on ci/taos/plugins-base/pr-audit-build-ubuntu.sh:
```diff
@@ -94,23 +95,22 @@ function pr-audit-build-ubuntu-run-queue(){
+        echo -e "[DEBUG] Starting 'pdebuild --use-pdebuild-internal ...' command."
+        echo -e "[DEBUG] The pdebuild start at $(date -R)"
+        time pdebuild --use-pdebuild-internal \
+        --logfile ${input_pr} --logfile ${input_date} \
```
It is a trick to figure out the victim process.

leemgs commented:
The local and unit test is completed by V6 finally. This PR is ready to be merged.
